File: bindings/python/src/anki_forge/options.py
from __future__ import annotations
from dataclasses import dataclass, replace, asdict
from enum import StrEnum
from os import PathLike
from pathlib import Path
from typing import Any
from . import _native
from ._bridge import invoke
PathInput = str | PathLike[str]

def absolute_path(path: PathInput, base_dir: Path | None = None) -> str:
    value = Path(path).expanduser()
    return str(value.absolute() if value.is_absolute() else ((base_dir or Path.cwd()) / value).absolute())

@dataclass(frozen=True)
class InspectLimits:
    max_archive_bytes: int = 2 << 30
    max_entries: int = 100_000
    max_central_directory_bytes: int = 16 << 20
    max_zip_entry_bytes: int = 1 << 30
    max_zip_total_bytes: int = 4 << 30
    max_meta_bytes: int = 64 << 10
    max_media_map_bytes: int = 16 << 20
    max_identity_bytes: int = 64 << 20
    max_collection_bytes: int = 512 << 20
    max_media_bytes: int = 256 << 20
    max_decoded_total_bytes: int = 4 << 30
    max_zstd_window_bytes: int = 64 << 20

class RiskLevel(StrEnum):
    INFO = "info"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

@dataclass(frozen=True)
class UpdatePolicy:
    _threshold: RiskLevel = RiskLevel.HIGH
    _allowances: tuple[str, ...] = ()

    def fail_on(self, level: RiskLevel) -> UpdatePolicy:
        return replace(self, _threshold=level)

    def allow(self, code: str) -> UpdatePolicy:
        invoke(_native.validate_risk_code, code)
        return replace(self, _allowances=(*self._allowances, code))

    def _payload(self) -> dict[str, Any]:
        return dict(threshold=self._threshold, allowances=self._allowances)

@dataclass(frozen=True)
class BuildOptions:
    _output: str | None
    _baseline: str | None = None
    _limits: InspectLimits | None = None
    _policy: UpdatePolicy | None = None

    @staticmethod
    def to(path: PathInput) -> BuildOptions:
        return BuildOptions(absolute_path(path))

    @staticmethod
    def temporary() -> BuildOptions:
        return BuildOptions(None)

    def update_from(self, path: PathInput) -> BuildOptions:
        return replace(self, _baseline=absolute_path(path))

    def inspect_limits(self, limits: InspectLimits) -> BuildOptions:
        return replace(self, _limits=limits)

    def update_policy(self, policy: UpdatePolicy) -> BuildOptions:
        return replace(self, _policy=policy)

    def _payload(self) -> dict[str, Any]:
        return dict(output=self._output, baseline=self._baseline,
                    limits=asdict(self._limits) if self._limits else None,
                    policy=self._policy._payload() if self._policy else None)

@dataclass(frozen=True)
class CompareOptions:
    _baseline: str
    _limits: InspectLimits | None = None
    _policy: UpdatePolicy | None = None

    @staticmethod
    def against(path: PathInput) -> CompareOptions:
        return CompareOptions(absolute_path(path))

    def inspect_limits(self, limits: InspectLimits) -> CompareOptions:
        return replace(self, _limits=limits)

    def update_policy(self, policy: UpdatePolicy) -> CompareOptions:
        return replace(self, _policy=policy)

    def _payload(self) -> dict[str, Any]:
        return dict(baseline=self._baseline, limits=asdict(self._limits) if self._limits else None,
                    policy=self._policy._payload() if self._policy else None)

File: bindings/python/native/src/options.rs
use crate::{core_error, mapped};
use ankiforge::{
    build::{json::PathSnapshot, InspectLimits},
    update::{CompareOptions, RiskLevel, UpdatePolicy},
    BuildOptions,
};
use pyo3::{exceptions::PyValueError, prelude::*};
use serde::Deserialize;
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BuildInput {
    output: Option<PathSnapshot>,
    baseline: Option<PathSnapshot>,
    limits: Option<InspectInput>,
    policy: Option<PolicyInput>,
}
impl BuildInput {
    pub fn options(self) -> PyResult<BuildOptions> {
        let mut b = self
            .output
            .map(PathSnapshot::into_path_buf)
            .map_or_else(BuildOptions::temporary, BuildOptions::to);
        if let Some(p) = self.baseline {
            b = b.update_from(p.into_path_buf());
        }
        if let Some(l) = self.limits {
            b = b.inspect_limits(l.limits());
        }
        if let Some(p) = self.policy {
            b = b.update_policy(p.policy()?);
        }
        Ok(b)
    }
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CompareInput {
    baseline: PathSnapshot,
    limits: Option<InspectInput>,
    policy: Option<PolicyInput>,
}
impl CompareInput {
    pub fn options(self) -> PyResult<CompareOptions> {
        let mut b = CompareOptions::against(self.baseline.into_path_buf());
        if let Some(l) = self.limits {
            b = b.inspect_limits(l.limits());
        }
        if let Some(p) = self.policy {
            b = b.update_policy(p.policy()?);
        }
        Ok(b)
    }
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct PolicyInput {
    threshold: String,
    allowances: Vec<String>,
}
impl PolicyInput {
    fn policy(self) -> PyResult<UpdatePolicy> {
        let level = match self.threshold.as_str() {
            "info" => RiskLevel::Info,
            "low" => RiskLevel::Low,
            "medium" => RiskLevel::Medium,
            "high" => RiskLevel::High,
            "critical" => RiskLevel::Critical,
            _ => return Err(PyValueError::new_err("invalid risk level")),
        };
        let mut p = UpdatePolicy::default().fail_on(level);
        for c in self.allowances {
            p = p.allow(
                c.parse::<ankiforge::update::RiskCode>()
                    .map_err(mapped!("policy"))?,
            );
        }
        Ok(p)
    }
}
#[derive(Default, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct InspectInput {
    max_archive_bytes: Option<u64>,
    max_entries: Option<u64>,
    max_central_directory_bytes: Option<u64>,
    max_zip_entry_bytes: Option<u64>,
    max_zip_total_bytes: Option<u64>,
    max_meta_bytes: Option<u64>,
    max_media_map_bytes: Option<u64>,
    max_identity_bytes: Option<u64>,
    max_collection_bytes: Option<u64>,
    max_media_bytes: Option<u64>,
    max_decoded_total_bytes: Option<u64>,
    max_zstd_window_bytes: Option<u64>,
}

impl InspectInput {
    pub fn limits(self) -> InspectLimits {
        let mut limits = InspectLimits::default();
        if let Some(value) = self.max_identity_bytes {
            limits.max_identity_bytes = value;
        }
        if let Some(value) = self.max_archive_bytes {
            limits.max_archive_bytes = value;
        }
        if let Some(value) = self.max_entries {
            limits.max_entries = value;
        }
        if let Some(value) = self.max_central_directory_bytes {
            limits.max_central_directory_bytes = value;
        }
        if let Some(value) = self.max_zip_entry_bytes {
            limits.max_zip_entry_bytes = value;
        }
        if let Some(value) = self.max_zip_total_bytes {
            limits.max_zip_total_bytes = value;
        }
        if let Some(value) = self.max_meta_bytes {
            limits.max_meta_bytes = value;
        }
        if let Some(value) = self.max_media_map_bytes {
            limits.max_media_map_bytes = value;
        }
        if let Some(value) = self.max_collection_bytes {
            limits.max_collection_bytes = value;
        }
        if let Some(value) = self.max_media_bytes {
            limits.max_media_bytes = value;
        }
        if let Some(value) = self.max_decoded_total_bytes {
            limits.max_decoded_total_bytes = value;
        }
        if let Some(value) = self.max_zstd_window_bytes {
            limits.max_zstd_window_bytes = value;
        }
        limits
    }
}
